src/processor/reconnection-graph.service.ts

```diff
@@ -430,13 +419,71 @@ export class ReconnectionGraphService {
     return dsnpKeys;
   }
 
-  async processChainEvents(promises: Promise<ParsedEventResult>[], dsnpUserId: MessageSourceId): Promise<void> {
-    // loop over promises and wait for all to resolve
-    for (const promise of promises) {
-      const [event, eventMap] = await promise;
+  async sendAndProcessChainEvents(
+    dsnpUserId: MessageSourceId,
+    providerId: ProviderId,
+    providerKeys: KeyringPair,
+    batchesMap: SubmittableExtrinsic<'rxjs', ISubmittableResult>[][],
+    ): Promise<void> {
+    try {
+      // iterate over batches and send them to the chain
+      batchesMap.forEach(async (batch, batchIndex) => {
+        await this.processSingleBatch(dsnpUserId, providerId, providerKeys, batch);
+      });
+    } catch (e) {
```

Comment:
An `async` forEach loop doesn't work the way you'd expect... you need to collect the promises and `await Promise.all()`

https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop